### DISCUSSION TITLE
pin memory에 대하여

### DISCUSSION THREAD

**Discussion by Grapefruitsaft** [bcaitech1/p4-dkt-DKTJHGSD#46] [Show and tell]
![image](https://user-images.githubusercontent.com/23133930/119809897-81a70f00-bf20-11eb-8b98-ee3cd3a92baf.png)
Device - GPU
Host - RAM + DISK

**pageable memory**
- 잠시 사용하지 않는 메모리 부분을 page 단위로 디스크(SSD/HDD)에 swap 해주어 용량이 큰 프로그램을 올릴 수 있습니다.
- RAM 과 DISK간 데이터 이동이 존재합니다.
- 만약 DISK에 들어가 있는 데이터를 접근하려면 DISK에서 RAM으로 가져 오는 작업이 필요하고 이 과정에서 overhead가 발생하게 됩니다.
- 보통 OS에서 사용하는 user space memory는 pageable memory 입니다. 
(page 와 swap 개념은 OS에서 나오는 CS지식이니 여기에서는 서술하지 않겠습니다. 대신 관련 링크 첨부드립니다.)
https://en.wikipedia.org/wiki/Virtual_memory
https://en.wikipedia.org/wiki/Page_(computer_memory)

**pinned(page-locked) memory**
- swap이 일어나지 않는 메모리를 의미합니다. 
- 데이터는 RAM에서만 존재합니다.

GPU는 pageable memory에 직접 접근하지 못하기 때문에 만약 접근하려는 데이터가 pageable memory에 존재한다면 pageable memory에 있는 데이터를 pinned memory에 옮겨주어야 합니다.  
이 과정에서 pageable memory 에서 pinned memory로 카피하는 과정에서 오버헤드가 발생해 추가 시간이 걸립니다.

만약, 데이터를 처음부터 pinned memory에 할당해준다면 카피하는 과정이 없어지므로 오버헤드가 줄어 학습 시간을 절약할 수 있습니다.

그렇기에 `dataloader.py` 에 정의된 `get_loader`함수의 변수 `pin_memory`를 `True`로 설정한다면 데이터 로더를 pin memory에 할당해줄 수 있습니다. :) 

**👉   해주어야할 작업!**
`pin_memory = True` in get_loader,  line 178, dataloader.py

**reference**
https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/

**Comment by joochanhyeong** [comment]
오..........모르는 사람 엄청 많을 것 같아요. 자세한 조사 감사합니다!!

**Reply by junho621** [reply]
감사합니다!

**Comment by sooyounlee** [comment]
소정님 감사합니다!! :)
